folding-schemes/src/frontend/circom/utils.rs

```diff
@@ -0,0 +1,146 @@
+    pub fn extract_r1cs_and_witness(
+        &self,
+        inputs: &[(String, Vec<BigInt>)],
+    ) -> Result<(R1CS<F>, Option<Vec<F>>), Error> {
+        // Extracts the R1CS
+        let file = File::open(&self.r1cs_filepath)?;
+        let reader = BufReader::new(file);
+        let r1cs_file = r1cs_reader::R1CSFile::<F>::new(reader)?;
+        let r1cs = r1cs_reader::R1CS::<F>::from(r1cs_file);
+
+        // Extracts the witness vector
+        let witness_vec = self.extract_witness(inputs)?;
+
+        Ok((r1cs, Some(witness_vec)))
+    }
+
+    // Extracts the witness vector as a vector of PrimeField elements.
+    pub fn extract_witness(&self, inputs: &[(String, Vec<BigInt>)]) -> Result<Vec<F>, Error> {
+        let witness_bigint = self.calculate_witness(inputs)?;
+
+        witness_bigint
+            .iter()
+            .map(|big_int| {
+                self.num_bigint_to_ark_bigint(big_int)
+                    .and_then(|ark_big_int| {
+                        F::from_bigint(ark_big_int)
+                            .ok_or_else(|| Error::Other("could not get F from bigint".to_string()))
+                    })
+            })
+            .collect()
+    }
+
+    // Calculates the witness given the Wasm filepath and inputs.
+    pub fn calculate_witness(
+        &self,
+        inputs: &[(String, Vec<BigInt>)],
+    ) -> Result<Vec<BigInt>, Error> {
+        let mut calculator = WitnessCalculator::new(&self.wasm_filepath)
+            .map_err(|_| Error::Other("Could not create new WitnessCalculator".to_string()))?;
+        calculator
+            .calculate_witness(inputs.iter().cloned(), true)
+            .map_err(|_| Error::Other("Failed to calculate witness".to_string()))
+    }
+
+    // Converts a num_bigint::BigInt to a PrimeField::BigInt.
+    pub fn num_bigint_to_ark_bigint(&self, value: &BigInt) -> Result<F::BigInt, Error> {
+        let big_uint = value
+            .to_biguint()
+            .ok_or(Error::Other("BigInt is negative".to_string()))?;
+        F::BigInt::try_from(big_uint).map_err(|_| {
+            Error::Other("Failed to convert num_bigint::BigInt to PrimeField::BigInt".to_string())
+        })
+    }
+
+    // Converts a PrimeField element to a num_bigint::BigInt representation.
+    pub fn ark_primefield_to_num_bigint(&self, value: F) -> BigInt {
+        let primefield_bigint: F::BigInt = value.into_bigint();
+        let bytes = primefield_bigint.to_bytes_be();
+        BigInt::from_bytes_be(Sign::Plus, &bytes)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ark_bn254::Fr;
+    use ark_circom::circom::{CircomBuilder, CircomConfig};
+    use ark_circom::CircomCircuit;
+    use ark_relations::r1cs::{ConstraintSynthesizer, ConstraintSystem};
+    use std::path::PathBuf;
+
+    //To generate .r1cs and .wasm files, run the below command in the terminal.
+    //bash ./folding-schemes/src/frontend/circom/test_folder/compile.sh
```

Comment:
Maybe having a build script for this could be useful?

Reply:
As discussed, we (1) don't want to force people to install circom to use sonobe and (2) sonobe's circom support could change in the future. Hence, we might not want to have a build script for this.

Reply:
I was not sure what to do. Thanks for letting me know.